tree/20_cfe_basics/time_lib.cf

```diff
@@ -0,0 +1,203 @@
+#
+# If there is an error, a class is raised: job_${job_id}_error
+#    # Sample code to use in the caller 
+#      "job_${job_id}_after_splay" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.max_run_id}");
+#      "job_${job_id}_after_run" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+#      "job_${job_id}_run" expression => strcmp("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+bundle agent schedule_generic(job_id, agent_periodicity, splay, job_periodicity, minutes_from_start)
+{
+  vars:
+    any::
+      "test" string => "value";
+    noerror::
+      # shift between each run = run_periodicity % agent_periodicity
+      "run_shift_real" string => eval("${job_periodicity} % ${agent_periodicity}", "math", "infix");
+      "run_shift" int => "${run_shift_real}";
+
+      # max_run_id = int(splay / agent periodicity) = min number of agent run in interval
+      "max_run_id_real" string => eval("${splay}/${agent_periodicity}", "math", "infix");
+      "max_run_id" int => "${max_run_id_real}";
+
+      # run_id = int(splay / splay_max * max_run_id) = run id of agent at wich we must run the job
+      "job_run_id_real" string => eval("${splay.splay_dec} * ${max_run_id} / ${splay.splay_max}", "math", "infix");
+      "job_run_id" int => "${job_run_id_real}";
+
+      # elapsed in period = minutes_from_start % run periodicity = duration in minutes since the last start of the run period 
+      "elapsed" string => ifelse("null_job_periodicity", "${minutes_from_start}", eval("${minutes_from_start} % ${job_periodicity}", "math", "infix"));
+
+      # current_run_id = int(elapsed / agent_periodicity) = run id of current run
+      "current_run_id_real" string => eval("${elapsed} / ${agent_periodicity}", "math", "infix");
+      "current_run_id" int => "${current_run_id_real}";
+
+  classes:
+    any::
+      # if agent_periodicity = 0 -> error
+      "job_${job_id}_error_null_agent_periodicity" expression => strcmp("${agent_periodicity}", "0"),
+                                                   scope => "namespace";
+      # if run_periodicity = 0 
+      "null_job_periodicity" expression => strcmp("${job_periodicity}", "0");
+
+      # if agent_periodicity > splay -> warning
+      "splay_too_small" expression => islessthan("${splay}", "${agent_periodicity}");
+      "splay_null" expression => strcmp("${splay}", "0");
+
+      # if splay >= run_periodicity -> error
+      # this must be checked in parent too if you have multiple periodicity
+      "job_${job_id}_error_splay_too_big" not => islessthan("${splay}", "${job_periodicity}"),
+                                          scope => "namespace",
+                                          ifvarclass => "!null_job_periodicity";
+
+      # if run_shift != 0 -> unpredictable
+      "unpredictable" not => strcmp("${run_shift}", "0");
```

Comment:
run_shift is a the shift between each run on the same period, which means a non 0 values will produce a delay that is always increasing